soc/ti/k3/am6x/Kconfig

```diff
@@ -1,6 +1,9 @@
 # Copyright (c) 2023 Enphase Energy
 # SPDX-License-Identifier: Apache-2.0
 
+config SOC_SERIES_AM6X
+	select PINCTRL
```

Comment:
this is not correct, the **drivers** need to select it

Reply:
@nordicjm I am a bit confused here, since I think beagleplay and beagleconnect freedom are also currently doing this wrong. There is `select PINCTRL` in soc Kconfig: https://github.com/zephyrproject-rtos/zephyr/blob/4793e00ae803ae6b7948e4c110f7ceef11821ef5/soc/ti/simplelink/cc13x2x7_cc26x2x7/Kconfig#L18

From your comment it seems that drivers are supposed to do this. However, where should pinctrl be enabled in case the board initialization needs it ?(the antenna switch in bcf and play case: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/beagle/beagleconnect_freedom/board_antenna.c)

Reply:
We use the PINCTRL for the `uart_ns16550.c` driver, but that driver doesn't select it, should I go and fix that up everywhere?

Reply:
That won't work either, as some users of `uart_ns16550.c` need PINCTRL and others do not. We can only know at the board level if pinmux will be needed. Not sure I know a good solution.

Reply:
`select PINCTRL if <soc_name>`

Reply:
This would need added for every current K3 <soc_name>, and then updated for every new one added, sounds like constant churn in the serial driver kconfig. Having it selected by the symbol `UART_NS16550_TI_K3` which is in `Kconfig.ns16550` and already selected by all the K3 SoCs seems to have the same effect (all K3 SoCs have pinmux) without listing each and every SoC in each and every driver's kconfig.

Reply:
Except if a user disables the UART, you're forcing on an option that is not needed because your dependencies are not correct

Reply:
If the user disables `SERIAL` then `UART_NS16550_TI_K3` will also not be enabled, and so `PINCTRL` would not be selected.

Reply:
BTW I just wanted to note that I do agree with what you are saying here.

If we want to do the right thing in all cases, then one must also note that `PINCTRL` is really only needed on boards that don't use the default pinmux settings for the SoC (all SoC pins have a default mux setting, if the board uses those then `PINCTRL` would also not be needed). This can only be known at the board level. But also as you note, only some drivers support `PINCTRL` in the first place (all should at some point, or it should be moved to a higher level like in Linux and done for drivers automatically in the `default` case).

So we only select `PINCTRL` if 3 things are true, the board's DT has a pinmux node associated with a driver, that driver uses pinmux, and that driver is enabled. If any of these or not true for all drivers used by a board, only then can `PINCTRL` be not selected.

I'm thinking there should be some DT parsing magic that can help with this. For all driver DT nodes that contain an enabled pinmux property we define `DT_HAS_<DRIVER>_PINMUX_ENABLED`. Then in Kconfig for drivers we can do this:

```
config <DRIVER>
	depends on DT_HAS_<DRIVER>_ENABLED
	select PINCTRL if DT_HAS_<DRIVER>_PINMUX_ENABLED
```

Reply:
PINCTRL needs to be selected if the associated code makes use of the pinctrl API, that’s all you need to take care of.